src/rust/ide/src/ide/integration.rs

```diff
@@ -1252,16 +1177,126 @@ impl Model {
         })
     }
 
-    fn get_controller_visualization_id
-    (&self, node_id:graph_editor::NodeId) -> Result<VisualizationId,NoSuchVisualization> {
-        let err = || NoSuchVisualization(node_id);
-        self.visualizations.get_copied(&node_id).ok_or_else(err)
-    }
-
     fn lookup_computed_info(&self, id:&ExpressionId) -> Option<Rc<ComputedValueInfo>> {
         let registry = self.graph.computed_value_info_registry();
         registry.get(id)
     }
+
+    fn attach_visualization
+    ( &self
+    , node_id               : graph_editor::NodeId
+    , vis_metadata          : &Option<visualization::Metadata>
+    , receive_data_endpoint : frp::Any<(graph_editor::NodeId,visualization::Data)>
+    , visualizations_map    : SharedHashMap<graph_editor::NodeId,VisualizationId>
+    ) -> FallibleResult<VisualizationId> {
+        // Do nothing if there is already a visualization attached.
+        let err = || VisualizationAlreadyAttached(node_id);
+        (!visualizations_map.contains_key(&node_id)).ok_or_else(err)?;
+
+        debug!(self.logger, "Attaching visualization on {node_id}.");
```

Comment:
is `node_id` just a number? if so, I'd use here `"Attaching visualization on node {node_id}."` instead.